doc/developer/style.md

```diff
@@ -281,6 +281,38 @@ do not produce useful backtraces and instead just point to the line in libtest t
 that the test didn't return an error. Panics will produce useful backtraces that include
 the line where the panic occurred. This is especially useful for debugging flaky tests in CI.
 
+### Errors
+
+#### Prefer structured errors over strings
+
+You should prefer structured error enums over strings. We currently have a
+number of places where we use `anyhow!("my error message")`, but we are trying
+to migrate away from that. Using [`thiserror`] makes it easy to write
+structured error enums.
+
+If you write your own error enums, make sure to implement the standard
+[`Error`] trait and to properly implement `source()`. This makes sure that we
+can get the chain of causing/underlying errors so that they can be reported
+correctly when needed. With [`thiserror`] this will happen automatically when
+you use the `#[from]` attribute.
+
+Generally, the `Display` impl of your error type should _not_ print the chain
+of errors but only print itself and rely on callers to print the error chain
+when needed. There could be exceptions when your error type is wrapping another
+error type or is compositionally including one or multiple other errors, but
+those should be very rare!
+
+#### Printing errors
+
+As mentioned above, the `Display` impl of an error should not print the chain
+of source errors. Whenever you _do_ need to print an error with its chain of
```

Comment:
I'm confused, because afaik, we only ever trace/log errors or surface them to users. Are there other ways we print errors? When surfacing to users they all get converted into a hopefully structured SQL error, and (almost?) never have their chain included. I've read this paragraph a few times and don't understand when I'd want to use those functions still.

Reply:
I should have included more context here or in the PR description! This whole thing got started by https://github.com/MaterializeInc/materialize/issues/18490#issuecomment-1495385815.

As to where we print errors and where we use the `_with_causes` variants, I'll try and list the different "flavours" we have, with an example. The only real user facing errors should be (as you pointed out) the errors we report back via the SQL client/pgwire _and_ the errors that we report in the error/status history collections.
 
Errors that we report in `mz_source_status_history` (and related friends and versions for sinks): https://github.com/MaterializeInc/materialize/blob/2191ed501ab8ff868c0584317574de38c3be1881/src/storage/src/sink/kafka.rs#L907

Structured SQL errors which include a chain of errors in the _details_, or in the error itself:
https://github.com/MaterializeInc/materialize/blob/acc39ea1b9841fef17e42732754bc26e801a00ef/src/adapter/src/error.rs#L266
and
https://github.com/MaterializeInc/materialize/blob/86cc7805b61671768ec11460c6a59d6e82382ca6/src/sql/src/plan/error.rs#L180

A lot of persist errors, which we mostly log internally, or when panick'ing:
https://github.com/MaterializeInc/materialize/blob/1804102abf16e2c2ac99b3aaefc317883329f06b/src/persist-client/src/internal/compact.rs#L378

Top-level error handlers for the processes:
https://github.com/MaterializeInc/materialize/blob/22d35d575515d2a2383462fe7fda9df3ecf4c9e0/src/environmentd/src/bin/environmentd/main.rs#L532

Some that might be pathological cases, where we already have the chain of errors in the unstructured error itself:
https://github.com/MaterializeInc/materialize/blob/acc39ea1b9841fef17e42732754bc26e801a00ef/src/adapter/src/error.rs#L459
https://github.com/MaterializeInc/materialize/blob/86cc7805b61671768ec11460c6a59d6e82382ca6/src/sql/src/plan/error.rs#L466